omaha/core/scheduler_unittest.cc

```diff
@@ -0,0 +1,180 @@
+// Copyright 2007-2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ========================================================================
+
+#include "omaha/core/scheduler.h"
+
+#include "omaha/base/constants.h"
+#include "omaha/base/highres_timer-win32.h"
+#include "omaha/testing/unit_test.h"
+#include "omaha/third_party/smartany/scoped_any.h"
+
+namespace omaha {
+
+namespace {
+
+inline void ASSERT_SIGNALLED_BEFORE(HANDLE handle, DWORD timeout_ms) {
+  ASSERT_EQ(WAIT_OBJECT_0, ::WaitForSingleObject(handle, timeout_ms));
+}
+
+inline void ASSERT_ALL_SIGNALLED_BEFORE(int count,
+                                        HANDLE* handles,
+                                        DWORD timeout_ms) {
+  // Wait for all handles
+  constexpr bool kWaitAll = true;
+  ASSERT_EQ(WAIT_OBJECT_0,
+            ::WaitForMultipleObjects(count, handles, kWaitAll, timeout_ms));
+}
+
+inline void ASSERT_TIMEOUT_AFTER(HANDLE handle, DWORD timeout_ms) {
+  ASSERT_EQ(WAIT_TIMEOUT, ::WaitForSingleObject(handle, timeout_ms));
+}
+
+}  // namespace
+
+class SchedulerTest : public ::testing::Test {
+ protected:
+  void SetUp() override {}
+  void TearDown() override {}
+};
+
+TEST_F(SchedulerTest, ScheduledTaskReschedules) {
+  int call_count = 0;
+  std::vector<HANDLE> event_handles(4);

```

Comment:
Could we use scoped_event instead of raw handle? In this case, the handles are leaked, not a big deal, but then code that detects such conditions may trigger.

Reply:
Done